plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -212,17 +214,132 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
-            -- If we use the context-based approach like in GHC, this won't be a problem, so we may
-            -- consider that in the future.
-            inlineSaturatedApp =<< forMOf termSubterms t processTerm
+        t -> do
+            -- See note [Processing order of call site inlining]
+            let (hd, args) = splitApplication t
+                -- process the term args, ignore the type arguments
+                processArgs :: AppContext tyname name uni fun ann
+                    -> InlineM tyname name uni fun ann (AppContext tyname name uni fun ann)
+                processArgs (TermAppContext arg ann ctx) = do
+                    processedArg <- processTerm arg
+                    processedArgs <- processArgs ctx
+                    pure $ TermAppContext processedArg ann processedArgs
+                processArgs (TypeAppContext ty ann ctx) = do
+                    processedArgs <- processArgs ctx
+                    pure $ TypeAppContext ty ann processedArgs
+                processArgs AppContextEnd = pure AppContextEnd
+            case (hd, args) of
+                -- not an application, just process the subterms
+                (_, AppContextEnd) -> forMOf termSubterms t processTerm
+                (Var{}, _) -> do
+                    processedHd <- processTerm hd -- the variable may be unconditionally inlined.
+                    case processedHd of
+                        -- if it didn't get unconditionally inlined/got inlined to another variable
+                        (Var _ name) -> do
+                                gets (lookupVarInfo name) >>= \case
+                                    Just varInfo -> do
+                                        let defAsInlineTerm = varRhs varInfo
+                                            inlineTermToTerm :: InlineTerm tyname name uni fun ann
+                                                -> Term tyname name uni fun ann
+                                            inlineTermToTerm (Done (Dupable var)) = var
+                                            -- extract out the rhs without renaming, we only rename
+                                            -- when we know there's substitution
+                                            rhs = inlineTermToTerm defAsInlineTerm
+                                        -- process the args
+                                        processedArgs <- processArgs args
+                                        maybeInlined <- callSiteInline t rhs varInfo processedArgs
+                                        case maybeInlined of
+                                            Just inlined -> pure inlined
+                                            -- we didn't inline at the call site, just process the
+                                            -- subterms
+                                            Nothing      -> forMOf termSubterms t processTerm
```

Comment:
Should just be `pure t` here. The subterms have already been processed

Reply:
I don't think so? The processed args aren't returned when `callSiteInline` returns `Nothing`?

Reply:
I tried `pure t` and one of the tests (`factorial`) got a free variable error.

Reply:
Oh, not `pure t`, but `pure (fillAppContext processedHd processedArgs)`.